File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/agent-tool-capability-risk.ts
import { AGENT_TOOL_DANGEROUS_CAPABILITY_PATTERN } from "../../constants/security-scan.js";
import { defineRule } from "../../utils/define-rule.js";
import { isProductionSourcePath } from "./utils/is-production-source-path.js";
import { scanByPattern } from "./utils/scan-by-pattern.js";

const AGENT_TOOL_DEFINITION_PATTERN =
  /\b(?:tool\s*\(\s*\{|createTool\s*\(|defineTool\s*\(|new\s+(?:DynamicTool|StructuredTool)\s*\()/;

const AGENT_TOOL_CONTEXT_PATH_PATTERN =
  /(?:^|\/)(?:agents?|tools?|mcp)(?:\/|$)|(?:agent|tool|mcp)[^/]*\.[cm]?[jt]sx?$/i;

export const agentToolCapabilityRisk = defineRule({
  id: "agent-tool-capability-risk",
  title: "Agent tool exposes dangerous capability",
  severity: "warn",
  recommendation:
    "Treat tool inputs as prompt-injection controlled. Validate arguments, scope permissions per call, and avoid exposing shell/file/network primitives directly to agents.",
  scan: scanByPattern({
    shouldScan: (file) =>
      isProductionSourcePath(file.relativePath) &&
      AGENT_TOOL_CONTEXT_PATH_PATTERN.test(file.relativePath),
    pattern: AGENT_TOOL_DEFINITION_PATTERN,
    // In code only: the AI-SDK `execute:` handler key and capability words in
    // `description` prose ("...fetch the numbers...") must not satisfy the gate.
    requireAllInCode: [AGENT_TOOL_DANGEROUS_CAPABILITY_PATTERN],
    message:
      "An agent-callable tool appears to expose network, filesystem, shell, or code-execution capability.",
  }),
});

File: native/oxlint/scans/agent-tool-capability-risk.rs
use lazy_regex::{Lazy, Regex, lazy_regex};

use super::{
    ScanFinding, first_pattern_finding::first_pattern_finding, scan_content::ScanContent,
};

const MESSAGE: &str = "An agent-callable tool appears to expose network, filesystem, shell, or code-execution capability.";

static TOOL_DEFINITION_PATTERN: Lazy<Regex> = lazy_regex!(
    r"\b(?:tool\s*\(\s*\{|createTool\s*\(|defineTool\s*\(|new\s+(?:DynamicTool|StructuredTool)\s*\()"
);
static TOOL_PREFILTER_PATTERN: Lazy<Regex> =
    lazy_regex!(r"\b(?:tool|createTool|defineTool|DynamicTool|StructuredTool)\b");
static TOOL_CONTEXT_PATH_PATTERN: Lazy<Regex> =
    lazy_regex!(r"(?i)(?:^|/)(?:agents?|tools?|mcp)(?:/|$)|(?:agent|tool|mcp)[^/]*\.[cm]?[jt]sx?$");
static DANGEROUS_CAPABILITY_PATTERN: Lazy<Regex> = lazy_regex!(
    r"\b(?:exec|execSync|spawn|child_process|eval|new Function|vm\.run|readFile|writeFile|fs\.read|fs\.write|fetch|axios|http\.request|sandbox|runCode|executeCode)\b"
);

pub fn scan(relative_path: &str, source: &ScanContent<'_>) -> Vec<ScanFinding> {
    if !super::is_production_file_path::is_production_source_path(relative_path) {
        return Vec::new();
    }
    let normalized_path =
        super::normalize_js_regex_content::normalize_js_regex_content(relative_path);
    if !TOOL_CONTEXT_PATH_PATTERN.is_match(&normalized_path) {
        return Vec::new();
    }
    let normalized_source = source.normalized_source();
    if !TOOL_PREFILTER_PATTERN.is_match(&normalized_source)
        || !DANGEROUS_CAPABILITY_PATTERN.is_match(&normalized_source)
    {
        return Vec::new();
    }
    let content = source.normalized_scannable(true);
    if !DANGEROUS_CAPABILITY_PATTERN.is_match(&content) {
        return Vec::new();
    }
    first_pattern_finding(source, &content, &[&TOOL_DEFINITION_PATTERN], MESSAGE)
}
